### PR TITLE
Handle multiple codecs in renegotiation

### PR DESCRIPTION
**Pull request by cnderrauber**
update pion to v3.1.9 for answer same order of codec as publisher.
register enable codecs in subscriber peerconnection created.

### CODE REVIEW

**Review by boks1971 — APPROVED** [review]
Let get some H264 flowing :-)

@cnderrauber There is a place in sfu.DownTrack where we send canned 1x1 frame. I was using the bytes you generated via ffmpeg for VP8 for that. We will have to add support for H.264 there.

File: pkg/config/config.go [hunk 1]
```diff
@@ -146,15 +146,15 @@ func NewConfig(confString string, c *cli.Context) (*Config, error) {
 			ActiveLevel:     30, // -30dBov = 0.03
 			MinPercentile:   40,
 			UpdateInterval:  500,
-			SmoothIntervals: 4,
+			SmoothIntervals: 2,
```
Comment (cnderrauber) [inline pkg/config/config.go:149]: Test found that SmoothInterval set to 4 willl cause the speaker state reaction  a little slow. I have test set default to 2 will more sensitive. But don't know is there other reason to set SmoothIntervals to 4, if it has, can revert this change

**Review by davidzhao — CHANGES_REQUESTED** [review]
this looks great! awesome work on the audio level change too.

two asks:
1. could you split them into different PRs? this will make it easier to keep track of and revert if found problematic
2. could you add an integration test case for multiple codec types, it could work like this:
    * two clients in the room
    * client A publishes vp8, client B receives it
    * client A publishes h.264, client B receives it as well.

File: pkg/rtc/mediatrack.go [hunk 2]
```diff
@@ -403,7 +399,7 @@ func (t *MediaTrack) AddReceiver(receiver *webrtc.RTPReceiver, track *webrtc.Tra
 		t.simulcasted.TrySet(true)
 	}
 
-	buff.Bind(receiver.GetParameters(), buffer.Options{
+	buff.Bind(receiver.GetParameters(), track.Codec().RTPCodecCapability, buffer.Options{
```
Comment (cnderrauber) [inline pkg/rtc/mediatrack.go:402]: buff should use codec of TrackRemote it received

**Comment by cnderrauber** [comment]
> this looks great! awesome work on the audio level change too.
> 
> two asks:
> 
> 1. could you split them into different PRs? this will make it easier to keep track of and revert if found problematic
> 2. could you add an integration test case for multiple codec types, it could work like this:
>    
>    * two clients in the room
>    * client A publishes vp8, client B receives it
>    * client A publishes h.264, client B receives it as well.

Test case added. Will open another pr for audio level with dtx

File: pkg/sfu/buffer/buffer_test.go [hunk 3]
```diff
@@ -278,16 +268,8 @@ func TestFractionLostReport(t *testing.T) {
...
-			{
-				RTPCodecCapability: webrtc.RTPCodecCapability{
-					MimeType:  "audio/opus",
-					ClockRate: 48000,
-				},
-				PayloadType: 96,
-			},
-		},
-	}, Options{})
+		Codecs:           []webrtc.RTPCodecParameters{opusCodec},
```
Comment (boks1971) [inline pkg/sfu/buffer/buffer_test.go:271]: Nice cleanup!

File: pkg/sfu/downtrack.go [hunk 4]
```diff
@@ -1113,40 +1120,64 @@ func (d *DownTrack) writeBlankFrameRTP() error {
...
+
+	_, err = d.writeStream.WriteRTP(hdr, payload)
+	return err
+}
+
+func (d *DownTrack) writeH264BlankFrame(hdr *rtp.Header, frameEndNeeded bool) error {
+	// TODO - Jie Zeng
+	// now use STAP-A to compose sps, pps, idr together, most decoder support packetization-mode 1.
+	// if client only support packetization-mode 0, use single nalu unit packet
+	buf := make([]byte, 1462)
```
Comment (boks1971) [inline pkg/sfu/downtrack.go:1168]: boy, H.264 bitstream is complex. Thank you for putting this together.

File: test/singlenode_test.go [hunk 5]
```diff
@@ -139,3 +140,79 @@ func Test_WhenAutoSubscriptionDisabled_ClientShouldNotReceiveAnyPublishedTracks(
...
+				vp8Found = true
+			} else if strings.EqualFold(t.Codec().MimeType, "video/h264") {
+				h264Found = true
+			}
+		}
+		return vp8Found && h264Found
+	})
+	if !success {
+		t.FailNow()
+	}
```
Comment (boks1971) [inline test/singlenode_test.go:216]: Good one @cnderrauber

**Review by boks1971 — APPROVED** [review]
This is great Jie! Let's do some H.264.....

**Pull request merged**